In [76]:
from IPython.display import display, HTML
display(HTML("<style>.container { width:90% !important; }</style>"))
import numpy as np
import matplotlib.pyplot as plt

In [77]:
G = np.array([lambda x,y,z: 6*x-2*np.cos(y*z)-1,
     lambda x,y,z: 9*y + np.sqrt( x**2 + np.sin(z) + 1.06 ) + 0.9,
     lambda x,y,z: 60*z + 3*np.exp(-x*y)+10*np.pi - 3])
A= np.array([lambda x,y: np.log(y**2+x**2)-np.sin(x*y)-np.log(2)-np.log(np.pi),
     lambda x,y: np.exp(x-y)+np.cos(y*x)])

In [78]:
def GetF(G,r):
    
    n = r.shape[0]
    
    v = np.zeros_like(r)
    
    for i in range(n):
        if len(r)==3:
          v[i] = G[i](r[0],r[1],r[2])
        if len(r)==2:
          v[i] = G[i](r[0],r[1])  
        
    return v

In [79]:
def GetJacobian(f,r,h=1e-6):
    
    n = r.shape[0]
    
    J = np.zeros((n,n))
    
    for i in range(n):
        for j in range(n):
            
            rf = r.copy()
            rb = r.copy()
            
            
            rf[j] = rf[j] + h
            rb[j] = rb[j] - h
            if len(r)==3:
            
              J[i,j] = ( f[i](rf[0],rf[1],rf[2]) - f[i](rb[0],rb[1],rb[2])  )/(2*h)
            if len(r)==2:
               J[i,j] = ( f[i](rf[0],rf[1]) - f[i](rb[0],rb[1])  )/(2*h) 

            
    
    return J

In [80]:
def NewtonRaphson(G,r,itmax=1000,error=1e-9):
    
    it = 0
    d = 1.
    dvector = []
    
    while d > error and it < itmax:
        
        # Vector actual
        rc = r
        
        F = GetF(G,rc)
        J = GetJacobian(G,rc)
        InvJ = np.linalg.inv(J)
        
        r = rc - np.dot(InvJ,F)
        
        diff = r - rc
        
        d = np.max( np.abs(diff) )
        
        dvector.append(d)
        
        
        it += 1
    
    
    return r

In [81]:
r=NewtonRaphson(G,np.array([0.,0.,0.]))
r


array([ 0.49814468, -0.1996059 , -0.52882598])

Podemos concluir que la solución para este sistema de ecuaciones es x1=0.49814468,  x2=-0.1996059 x3=-0.52882598

In [82]:
a=NewtonRaphson(A, np.array([2.0,2.0]))
a

array([1.77245385, 1.77245385])

y para este sistema de ecuaciones las soluciones son x=y=1.77245385  